subsys/bluetooth/host/audio/iso.c

```diff
@@ -0,0 +1,1075 @@
+/*  Bluetooth ISO */
+
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+
+#include <bluetooth/hci.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+
+#include "host/hci_core.h"
+#include "host/conn_internal.h"
+#include "iso_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_AUDIO_DEBUG_ISO)
+#define LOG_MODULE_NAME bt_iso
+#include "common/log.h"
+
+NET_BUF_POOL_FIXED_DEFINE(iso_tx_pool, CONFIG_BT_ISO_TX_BUF_COUNT,
+			  CONFIG_BT_ISO_TX_MTU, NULL);
+NET_BUF_POOL_FIXED_DEFINE(iso_rx_pool, CONFIG_BT_ISO_RX_BUF_COUNT,
+			  CONFIG_BT_ISO_RX_MTU, NULL);
+
+#if CONFIG_BT_ISO_TX_FRAG_COUNT > 0
+NET_BUF_POOL_FIXED_DEFINE(iso_frag_pool, CONFIG_BT_ISO_TX_FRAG_COUNT,
+			  CONFIG_BT_ISO_TX_MTU, NULL);
+#endif /* CONFIG_BT_ISO_TX_FRAG_COUNT > 0 */
+
+/* TODO: Allow more than one server? */
+static struct bt_iso_server *iso_server;
+struct bt_conn iso_conns[CONFIG_BT_MAX_ISO_CONN];
+
+/* Audio Data Path direction */
+enum {
+	BT_ISO_CHAN_HOST_TO_CTRL,
+	BT_ISO_CHAN_CTRL_TO_HOST,
+};
+
+struct bt_iso_data_path {
+	/* Data Path direction */
+	uint8_t dir;
+	/* Data Path ID */
+	uint8_t pid;
+	/* Data Path param reference */
+	struct bt_iso_chan_path *path;
+};
+
+
+struct net_buf *bt_iso_get_rx(k_timeout_t timeout)
+{
+	struct net_buf *buf = net_buf_alloc(&iso_rx_pool, timeout);
+
+	if (buf) {
+		net_buf_reserve(buf, BT_BUF_RESERVE);
+		bt_buf_set_type(buf, BT_BUF_ISO_IN);
+	}
+
+	return buf;
+}
+
+void hci_iso(struct net_buf *buf)
+{
+	struct bt_hci_iso_hdr *hdr;
+	uint16_t handle, len;
+	struct bt_conn *conn;
+	uint8_t flags;
+
+	BT_DBG("buf %p", buf);
+
+	BT_ASSERT(buf->len >= sizeof(*hdr));
+
+	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+	len = sys_le16_to_cpu(hdr->len);
+	handle = sys_le16_to_cpu(hdr->handle);
+	flags = bt_iso_flags(handle);
+
+	iso(buf)->handle = bt_iso_handle(handle);
+	iso(buf)->index = BT_CONN_INDEX_INVALID;
+
+	BT_DBG("handle %u len %u flags %u", iso(buf)->handle, len, flags);
+
+	if (buf->len != len) {
+		BT_ERR("ISO data length mismatch (%u != %u)", buf->len, len);
+		net_buf_unref(buf);
+		return;
+	}
+
+	conn = bt_conn_lookup_handle(iso(buf)->handle);
+	if (!conn) {
+		BT_ERR("Unable to find conn for handle %u", iso(buf)->handle);
+		net_buf_unref(buf);
+		return;
+	}
+
+	iso(buf)->index = bt_conn_index(conn);
+
+	bt_conn_recv(conn, buf, flags);
+	bt_conn_unref(conn);
+}
+
+void hci_le_cis_estabilished(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_cis_established *evt = (void *)buf->data;
```

Comment:
Shouldn't the cast of the buf->data be of type `struct bt_hci_evt_le_cis_established *` instead of `void *`?

Reply:
I would guess void * is fine here since the equivalent net_buf_pull also would result in void * being returned.